### PR TITLE
Add reverse for vectors.

### PR DESCRIPTION
**Pull request by thelseraphim**
#### Submisison Checklist

- [ ] Run unit tests: `./runTests.py test/unit`
- [ ] Run cpplint: `make cpplint`
- [ ] Declare copyright holder and open-source license: see below

#### Summary:
Just a quick resolution of https://github.com/stan-dev/math/issues/344 adding a reverse utility for std::vectors. Like our sort functions, it mutates its argument. We ok with that?

#### Intended Effect:
Add a missing minor feature.

#### How to Verify:
Code is short and there's a unit test.

#### Side Effects:
None, adds a function.

#### Documentation:
doxygen

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by bob-carpenter** [comment]
Three things:

1.  It needs to be a constant reference, which is how arguments are passed in from the Stan language.  So it needs a copy.  

2.  It also needs tests for reverse, forward, and mixed modes.  That'll just test that the appropriate comparisons are defined for `std::reverse` to apply to the iterators.

3.  It should be implemented for row vectors and column vector types, too.  If you take an ordered vector, it's an Eigen vector, and that's the kind of thing we'll have users reversing.

**Comment by thelseraphim** [comment]
@bob-carpenter I hopefully fixed the const issue and am proceeding to the rest, I'm curious whether there are any good examples of of functions from vector to vector already in the language? I'm looking at `fun/sum` (in prim, rev, fwd), the autodiff paper, and a few other places but if there's an idiom it'd be nice to match

Also: I'm intending to put a little bit of noise on this pull request. Is there a nice way to avoid triggering CI builds on intermediate pushes?

**Comment by syclik** [comment]
In your commit, you should be able to include "[skip ci]", but I don't know how well that actually works as documented. 

Both only build on push, so you can prevent constant builds if you push when it's actually ready to test. 

Although, you really shouldn't worry about the builds that are fired off. Most of the time we're idle.

**Comment by jgabry** [comment]
[ci skip] in a commit message has always worked as advertised for me. It
definitely comes in handy.

**Comment by seantalts** [comment]
I've just been killing off the builds that I know I've triggered if I have updates to the PR that makes running those tests obsolete.

**Comment by bob-carpenter** [comment]
You also don't need to create a pull request until it's
ready to go.  We can review branches until then.  But
sometimes it's easier with pull requests.

- Bob

**Comment by bob-carpenter** [comment]
You can find the signatures in the math library functions
in the index of the manual.  So you're looking for something
that returns real[] or vector and takes a real[] or vector
argument.  For example, cumulative_sum.  But cumulative_sum
doesn't have specialized derivatives, so you're better off
looking at stan/math/rev/mat/fun/softmax.hpp

Also, we have an array allocator in memalloc_ that should've
been used in softmax.  You could replace that, too

    vari** alpha_vi_array
        = reinterpret_cast<vari**>(ChainableStack::memalloc_
                                   .alloc(sizeof(vari*) * alpha.size()));

should be

      vari** alpha_vi_array
        = ChainableStack::memalloc_.alloc_array<vari*>(alpha.size());

At least I think I have my types in a row here.  The memaalloc_ object
with alloc_array method is defined in stan/math/memory/stack_alloc.hpp.  

- Bob

**Comment by bob-carpenter** [comment]
jenkins, retest this please

**Comment by bob-carpenter** [comment]
I shouldn't have just kicked off retesting.  Looks like the reverse file isn't including the header for `std::begin` and `std::end`.

**Comment by thelseraphim** [comment]
@bob-carpenter it wasn't using it because I shouldn't have had it there. pushing a micro-change this time just to see the [ci skip] in action

File: test/unit/math/prim/arr/fun/reverse_test.cpp [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+TEST(MathFunctions, reverse) {
+  std::vector<double> x(3);
+  x[0] = 9;
+  x[1] = 2.2;
+  x[2] = 3.3;
+  std::vector<double> y = stan::math::reverse(x);
+  EXPECT_FLOAT_EQ(y[0], 3.3);
+  EXPECT_FLOAT_EQ(y[1], 2.2);
+  EXPECT_FLOAT_EQ(y[2], 9);
+}
```
Comment (bob-carpenter) [inline test/unit/math/prim/arr/fun/reverse_test.cpp]: The reverse function looks good now.   But the pull request still needs rev and fwd autodiff tests just to make sure that everything needed by std::reverse_copy is available for the autodiff classes.

**Comment by bob-carpenter** [comment]
Was this all ready for final review?

**Comment by bob-carpenter** [comment]
@thelseraphim  Would you mind adding the autodiff tests for this?  Just one for fwd, one for rev, and one for rev nested in fwd (three test files in rev, fwd, and mix directories).

**Comment by syclik** [comment]
@thelseraphim mind finishing out this pull request?

**Pull request closed without merging**